cranelift/codegen/src/isa/x64/lower.isle

```diff
@@ -3491,23 +3491,150 @@
+;; it needs a special case here since the instruction being crafted has a
+;; different shape.
+(rule 2 (lower (udiv a @ (value_type $I8) b))
+        (x64_div8 (extend_to_gpr a $I32 (ExtendKind.Zero))
+                  (nonzero_divisor $I8 b)
+                  $false))
+
+;; 16-to-64-bit division is all done with a similar instruction and the only
+;; tricky requirement here is that when div traps are disallowed the divisor
+;; must not be zero.
+(rule 1 (lower (udiv a @ (value_type (fits_in_64 ty)) b))
+        (x64_div_quotient a
+                          (imm $I64 0)
+                          (nonzero_divisor ty b)
+                          (raw_operand_size_of_type ty)
+                          $false))
+
+;; Helper to place `Value` into a `Gpr` while possibly trapping if it's zero.
+;;
+;; If the `avoid_div_traps=true` codegen setting is specified then the value
+;; is checked for zero and a trap happens before the value is returned as a
+;; register here.
+(decl nonzero_divisor (Type Value) Gpr)
+
+;; As a special-case if the divisor is a constant number which is nonzero then
+;; no matter what there's no checks necessary.
+(rule 2 (nonzero_divisor ty (iconst (u64_from_imm64 (u64_nonzero n))))
+        (imm ty n))
+
+;; No checks necessary when `avoid_div_traps=false`
+(rule 1 (nonzero_divisor ty val)
+        (if-let $false (avoid_div_traps))
+        val)
+
+;; Base case executes a `test` instruction and trap if it's equal to itself.
+(rule (nonzero_divisor ty val)
+      (let (
+          (val Reg val)
+          (_ InstOutput (side_effect (with_flags_side_effect
+            (x64_test (raw_operand_size_of_type ty) val val)
+            (trap_if (CC.Z) (TrapCode.IntegerDivisionByZero)))))
+        )
+        val))
 
 ;; Rules for `sdiv` ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
-(rule (lower (sdiv a @ (value_type ty) b))
-      (div_or_rem (DivOrRemKind.SignedDiv) a b))
+(rule 2 (lower (sdiv a @ (value_type $I8) b))
+        (let (
+            (a Gpr (x64_sign_extend_data a (OperandSize.Size8)))
+          )
+          (x64_div8 a (safe_sdiv_divisor $I8 a b) $true)))
+
+(rule 1 (lower (sdiv a @ (value_type (fits_in_64 ty)) b))
+        (let (
+            (a Gpr a)
+            (size OperandSize (raw_operand_size_of_type ty))
+            (b Gpr (safe_sdiv_divisor ty a b))
+          )
+          (x64_div_quotient a (x64_sign_extend_data a size) b size $true)))
+
+;; Similar to `nonzero_divisor` except this checks to make sure that the divisor
+;; provided as a `Value` is safe to divide into the dividend `Gpr` provided.
+(decl safe_sdiv_divisor (Type Gpr Value) Reg)
+
+;; If the divisor is a constant that isn't 0 or -1, then it's always safe so
+;; materialize it into a register.
+(rule 3 (safe_sdiv_divisor ty a (iconst imm))
+        (if-let n (safe_divisor_from_imm64 ty imm))
+        (imm ty n))
+
+;; With `avoid_div_traps=false` the divisor can be plumbed through.
```

Comment:
Let's add a comment here why: on x64, INT_MIN / -1 will cause a hardware trap, and this is a legal way to signal the out-of-bounds error with `avoid_div_traps=false`.